src/fhirtypes/ElementDefinition.ts

```diff
@@ -1162,6 +1162,15 @@ export class ElementDefinition {
         break;
       case 'Reference':
         value = value as FshReference;
+        if (value.sdType) {
+          let lineage: Metadata[] = [];
+          this.type.forEach(t =>
+            t.targetProfile.map(tp => (lineage = lineage.concat(this.getTypeLineage(tp, fisher))))
+          );
+          if (!lineage.find(md => md.sdType === (value as FshReference).sdType)) {
+            throw new InvalidTypeError(`Reference(${value.sdType})`, this.type);
+          }
+        }
```

Comment:
I'm not sure this is handled correctly when it comes to inheritance. Easiest to explain with an example:
```
Instance: MyObservation
InstanceOf: Observation
* subject = Reference(MyResource)

Instance: MyResource
InstanceOf: Resource
* id = "hmm"
```
The FSH above will not get any `InvalidTypeError`s in my testing, but I think it should. The `subject` field is supposed to have a Reference to a Patient, Group, Device, or Location. Resource is a parent of these four, not a child, so shouldn't it be an invalid type here? 

I honestly may just be overly confusing myself, so definitely double check me on this. But it seems like the type checking is almost happening backwards here. I think we would want to get the lineage of the `value.sdType`. Then if anything in the lineage of that `sdType` matches something in `this.type`, then we are good, because that means the `value.sdType` is, or is a child of, something in `this.type`.

Reply:
Hm this is a good question. I sort of was thinking that the example of `* subject = Reference(MyResource)` would be allowed. I talked to @ngfreiter about this and we're still not sure of what is allowed in FHIR. @cmoesel do you know if this should be allowed?

Reply:
I can see why you might think an instance of `Resource` would be fine, since `Patient`/`Group`/`Device` are all `Resource`s and any data in a `Resource` instance would be valid in a `Patient`/`Group`/`Devices` instances.  But... there are cases where I think this could be problematic.

For example, imagine that `subject` could only be a `Group` (maybe we profiled it that way).  A `Group` is a `Resource`, but `Group` _requires_ `1..1 type` and `1..1 actual` -- so an instance of `Resource` is _not_ a valid instance of `Group` (since `Resource` does not have `type` or `actual` fields).

Similarly, `Resource` has a few fields that _could_ be profiled out by a sub-type.  So what if a sub-type profiled out `language` to be `0..0`?  If something was supposed to be a reference to that sub-type, then some references to `Resource` would be invalid instances of the sub-type (if they had values for `language`).

So... I think Nick is probably on the right path about getting the lineage of `value.sdType` and checking if anything in there is a match for the declared type of the field.

Reply:
Okay that makes sense! That's what we were going back and forth on. In that case, I'll update to support this and add a test to check that it works!